galite-core/src/main/kotlin/org/kopi/galite/visual/VMenuTree.kt

```diff
@@ -18,8 +18,432 @@
+   */
+  private fun fetchModules(isUnicode: Boolean): MutableList<Module> {
+    TODO()
+  }
+
+  private fun fetchGroupRightsByUserId(modules: List<Module>) {
+    fetchRights(modules, SELECT_GROUP_RIGHTS_BY_USERID)
+  }
+
+  private fun fetchGroupRightsByGroupId(modules: List<Module>) {
+    fetchRights(modules, SELECT_GROUP_RIGHTS_BY_GROUPID)
+  }
+
+  private fun fetchUserRights(modules: List<Module>) {
+    fetchRights(modules, SELECT_USER_RIGHTS)
+  }
+
+  private fun fetchRights(modules: List<Module>, queryText: String) {
     TODO()
   }
+
+  private fun findModuleById(modules: List<Module>, id: Int): Module? {
+    for (i in modules.indices) {
+      val module = modules[i]
+      if (module.id == id) {
+        return module
+      }
+    }
+    return null
+  }
+
+  /**
+   * Fetches the favorites from the database.
+   */
+  private fun fetchFavorites() {
+    TODO()
+  }
+
+  /*
+   * Loads the accessible modules.
+   */
+  private fun loadModules(loadFavorites: Boolean): Array<Module> {
+    var localModules: MutableList<Module> = ArrayList()
+
+    transaction {
+      localModules = fetchModules(ApplicationConfiguration.getConfiguration()!!.isUnicodeDatabase())
+      if (groupName != null) {
+        fetchGroupRightsByGroupId(localModules)
+      } else {
+        fetchGroupRightsByUserId(localModules)
+        fetchUserRights(localModules)
+      }
+      if (loadFavorites) {
+        fetchFavorites()
+      }
+    }
+
+    if (!isSuperUser) {
+      // walk downwards because we remove elements
+      val iterator = localModules.listIterator(localModules.size - 1)
+      while (iterator.hasPrevious()) {
+        val module = iterator.previous()
+
+        // remove all modules where access is explicitly denied
+        if (module.accessibility == Module.ACS_FALSE) {
+          iterator.remove()
+        }
+      }
+    }
+    // add default modules
+    addLogoutModule(localModules)
+    // order the menus alphabetically
+    localizeModules(ApplicationContext.getDefaultLocale())
+    localModules.sort()
+    moduleArray = localModules.toTypedArray()
+    return moduleArray
+  }
+
+  /**
+   * Adds the default logout module
+   */
+  protected fun addLogoutModule(localModules: MutableList<Module>) {
+    val logout = Module(Int.MAX_VALUE,
+            USER_MENU,
+            "logout",
+            RootMenu.ROOT_MENU_LOCALIZATION_RESOURCE,
+            LogoutModule::class.java.name,
+            Module.ACS_TRUE, Int.MIN_VALUE,
+            null)
+    items.add(logout)
+    localModules.add(logout)
+  }
+
+  /**
+   * Sets the title of the frame
+   */
+  override fun setTitle(s: String?) {
+    if (s != null) {
+      if (s.contains(VlibProperties.getString("program_menu"))) {
+        super.setTitle(s)
+      } else {
+        super.setTitle(s + " - " + VlibProperties.getString("program_menu"))
+      }
+    } else {
+      super.setTitle(VlibProperties.getString("program_menu"))
+    }
+  }
```

Comment:
can we change it to when

Reply:
I guess it's better to keep it like this because we have simple if..else statements